satellite.go

```diff
@@ -0,0 +1,214 @@
+	SatelliteServerScriptUrl string `json:"satellite_server_script_url"`
+	HostConfigured           bool   `json:"host_configured"`
+	format                   string
+}
+
+// Error implement error interface for structure ConfigureSatelliteResult
+func (configureSatelliteResult ConfigureSatelliteResult) Error() string {
+	var result string
+	switch configureSatelliteResult.format {
+	case "json":
+		data, err := json.MarshalIndent(configureSatelliteResult, "", "    ")
+		if err != nil {
+			return err.Error()
+		}
+		result = string(data)
+	case "":
+		break
+	default:
+		result = "error: unsupported document format: " + configureSatelliteResult.format
+	}
+	return result
+}
+
+// pingSatelliteServer tries to ping Satellite server to be sure that user
+// tries to reach get and run bootstrap script from Satellite server.
+// We use following endpoint that is available for unauthenticated users:
+// https://satellite.sat.engineering.redhat.com/apidoc/v2/ping/ping.en.html
+// It is not necessary to try to reach this endpoint, but download and run
+// some bash script downloaded from the internet. It is risky action. For
+// this reason we do this extra step to minimize risk that some non-intentional
+// code is downloaded and run as root user.
+//
+// In the future we could use following endpoint to determine version of
+// Satellite (Katello) and
+// https://satellite.sat.engineering.redhat.com/apidoc/v2/ping/server_status.en.html
+func pingSatelliteServer(satelliteScriptUrl *url.URL, s *spinner.Spinner) error {
+	if s != nil {
+		s.Suffix = fmt.Sprintf(" Connecting Satellite server: %v", satelliteScriptUrl.Host)
+	}
+
+	// Copy URL from struct URL and modify scheme, path and query
+	satellitePingUrl := *satelliteScriptUrl
+	satellitePingUrl.Scheme = "https"
+	satellitePingUrl.Path = "/api/ping"
+	satellitePingUrl.RawQuery = ""
+
+	// We have to use insecure HTTPs connection, because most of the customers use
+	// self-signed certificates
+	tlsConfig := tls.Config{
+		InsecureSkipVerify: true,
+	}
+	transport := http.DefaultTransport.(*http.Transport).Clone()
+	transport.TLSClientConfig = &tlsConfig
+	client := &http.Client{Transport: transport}
+
+	response, err := client.Get(satellitePingUrl.String())
+	if err != nil {
+		return fmt.Errorf("ping satellite server failed with error: %v", err)
+	}
+	if response.StatusCode != 200 {
+		return fmt.Errorf("ping satellite server failed with status code %d", response.StatusCode)
+	}
+	defer response.Body.Close()
+	resBody, err := io.ReadAll(response.Body)
+	if err != nil {
+		return fmt.Errorf("unable to read body of ping satellite server response: %v", err)
+	}
+
+	satellitePingResponse := SatellitePingResponse{}
+	err = json.Unmarshal(resBody, &satellitePingResponse)
+	if err != nil {
+		return fmt.Errorf("unable to parse ping satellite server response: %v", err)
+	}
+
+	// In theory, we could check for the status of candlepin and candlepin auth, etc.,
```

Comment:
Is it safe to assume that if the HTTP server responds with 200, that's an
appropriate response to a "ping" request?

Reply:
1) I can delete this commented block of code.
2) In theory we can also remove `json.Unmarshal()` of the body, but as I already said, it is safer to do more checks that the server is really Satellite server than do less checks, because we run downloaded script as a root.

Reply:
If user provide full URL to some no-satellite script and there will be also REST API endpoint `/api/ping` (not so unlikely), then the script will be downloaded and run as a root user. For this reason I would rather check that returned object of `/api/ping` is really response from Satellite.

Reply:
That makes sense to me. Considering the risk we're taking here (downloading a
shell script and running it as root), we should do as much as we can to ensure
that this script is coming from a trusted source.